File: LeetCode_645/main.cpp
/**
    LeetCode_645
    https://leetcode.com/problems/set-mismatch/description/
    MatrixDeity, 2018.
*/

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> findErrorNums(vector<int>& nums) {
        int n = nums.size();
        for (int i = 0; i < n; ++i) {
            while (nums[i] != nums[nums[i] - 1]) {
                swap(nums[i], nums[nums[i] - 1]);
            }
        }
        for (int i = 0; i < n; ++i) {
            if (nums[i] != i + 1) {
                return { nums[i], i + 1 };
            }
        }
        return { -1, -1 };
    }
};

int main() {
    return 0;
}

File: LeetCode_645/main.go
/**
  LeetCode_645
  https: //leetcode.com/problems/set-mismatch/
  MatrixDeity, 2022.
*/

package leetcode645

type nothing struct{}

func findErrorNums(nums []int) []int {
	set := map[int]nothing{}
	sum := 0
	bad, good := -1, -1
	for _, num := range nums {
		sum += num
		if _, ok := set[num]; ok {
			bad = num
		}
		set[num] = nothing{}
	}

	badSum := sum - bad
	n := int64(len(nums))
	var rightSum int64
	if n%2 == 0 {
		rightSum = n / 2 * (n + 1)
	} else {
		rightSum = (n + 1) / 2 * n
	}

	good = int(rightSum) - badSum

	return []int{bad, good}
}
